build/mongodl.py

```diff
@@ -0,0 +1,665 @@
+    db.execute('DROP TABLE IF EXISTS components')
+    db.execute('DROP TABLE IF EXISTS downloads')
+    db.execute('DROP TABLE IF EXISTS versions')
+    db.execute(r'''
+        CREATE TABLE versions (
+            version_id INTEGER PRIMARY KEY,
+            date TEXT NOT NULL,
+            version TEXT NOT NULL,
+            githash TEXT NOT NULL
+        )
+    ''')
+    db.execute(r'''
+        CREATE TABLE downloads (
+            download_id INTEGER PRIMARY KEY,
+            version_id INTEGER NOT NULL REFERENCES versions,
+            target TEXT NOT NULL,
+            arch TEXT NOT NULL,
+            edition TEXT NOT NULL,
+            ar_url TEST NOT NULL,
+            data TEXT NOT NULL
+        )
+    ''')
+    db.execute(r'''
+        CREATE TABLE components (
+            component_id INTEGER PRIMARY KEY,
+            key TEXT NOT NULL,
+            download_id INTEGER NOT NULL REFERENCES downloads,
+            data TEXT NOT NULL,
+            UNIQUE(key, download_id)
+        )
+    ''')
+    with json_file.open('r') as f:
+        data = json.load(f)
+    for ver in data['versions']:
+        version = ver['version']
+        githash = ver['githash']
+        date = ver['date']
+        db.execute(
+            r'''
+            INSERT INTO versions (date, version, githash)
+            VALUES (?, ?, ?)
+            ''',
+            (date, version, githash),
+        )
+        version_id = db.lastrowid
+        for dl in ver['downloads']:
+            arch = dl.get('arch', 'null')
+            target = dl.get('target', 'null')
+            edition = dl['edition']
+            ar_url = dl['archive']['url']
+            db.execute(
+                r'''
+                INSERT INTO downloads (version_id, target, arch, edition, ar_url, data)
+                VALUES (?, ?, ?, ?, ?, ?)
+                ''',
+                (version_id, target, arch, edition, ar_url, json.dumps(dl)),
+            )
+            dl_id = db.lastrowid
+            for key, data in dl.items():
+                if 'url' not in data:
+                    continue
+                db.execute(
+                    r'''
+                    INSERT INTO components (key, download_id, data)
+                    VALUES (?, ?, ?)
+                    ''',
+                    (key, dl_id, json.dumps(data)),
+                )
+
+
+def get_dl_db():
+    caches = cache_dir()
+    caches.mkdir(exist_ok=True, parents=True)
+    db = sqlite3.connect(str(caches / 'downloads.db'), isolation_level=None)
+    db.executescript(r'''
+        CREATE TABLE IF NOT EXISTS meta (
+            etag TEXT,
+            last_modified TEXT
+        )
+    ''')
+    db.executescript(r'''
+        CREATE TABLE IF NOT EXISTS past_downloads (
+            url TEXT NOT NULL UNIQUE,
+            etag TEXT,
+            last_modified TEXT
+        )
+    ''')
+    changed, full_json = _download_file(
+        db, 'https://downloads.mongodb.org/full.json')
```

Comment:
Sorry, I'm not really sure what you mean in this comment.

As for "latest" URLs: They are _not_ contained within `full.json` (unless you do a scan of the versions and pick the highest version yourself). Except for one-off quick testing, I'm wary of any automated system that automatically pulls a moving target as a dependency.

Reply:
Since this script is entirely based around `full.json`, I'll postpone adding "latest" options, although it may be useful in the future. It would be handy if Evergreen posted a machine-parseable listing (Does it?)

Reply:
SGTM. I am not aware of machine-parseable listing of latest builds.